Style: no empty newlines at start or end of functions

tkelman opened the pull request:
These were the few outliers style-wise that looked like this. Mostly from `@carlobaldassi`.

tkelman committed:
Style: no empty newlines at start or end of functions

amitmurthy commented:
Can we formalize this with an appropriate addition to https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#general-formatting-guidelines-for-julia-code-contributions

nalimilan commented:
+1, we need to make the style guide more detailed.

tkelman commented:
Sure, if there's agreement on this point then it's just a matter of how to phrase the guideline.

amitmurthy commented:
Since source files are being changed to follow this style, may as well make it a guideline. I am fine with it.

nalimilan commented:
See https://github.com/JuliaLang/julia/pull/15439.

andreasnoack commented:
I think it can be useful for readability to have a newline in the beginning of a function when the signature is long. Is there a problem with allowing this? E.g. in terms of readability of diffs or something similar? Or is it mainly a matter of taste?

tkelman commented:
Consistency. And it wastes space in a way that doesn't clarify anything in my opinion.

tbreloff commented:
Personally, I like a space when the function can be broken up cleanly, or if there's a comment on the first line:

```julia
function nospace()
  dosomethingshort()
  dosomethingshort()
end

function yesspace()

  # first do this
  this()

  # then that
  that()

  # finally other
  other()

end
```

tkelman commented:
This isn't unique to function definitions, any block opener/closer where you would typically indent the contents, having an empty newline right after the block opener or right before the `end` is redundant with the indentation and visually decouples the start and end of the block from its contents. It sticks out and doesn't look right to me. I'm all for whitespace between multiple steps, but before-start or after-finished don't count - that's what the `function` and the `end` are for.

amitmurthy commented:
I concur with Tony on the consistency part, not just in a single file, but with rest of the codebase. Formalizing this as a preferred guideline (and not as a rule) would help. As Andreas points out, exceptions to the guideline could very well be long function signatures which wrap lines, where a newline would help with readability.

nalimilan commented:
+1 for consistency as well.

tkelman committed:
replace `= begin` with `function`

remove more extraneous newlines, rearrange signatures for VersionNumber

tkelman committed:
More style adjustments, some reindenting and removing excess newlines

Adjust some signatures for uniformity and line length

samoconnor commented on base/bitarray.jl:
```diff
@@ -600,7 +600,6 @@ function insert!(B::BitVector, i::Integer, item)
 
 function _deleteat!(B::BitVector, i::Integer)
-
     k, j = get_chunks_id(i)
```
Please don't delete blank lines in cases like this where the author has clearly taken some care to use whitespace to delineate functional blocks.

tkelman replied:
This makes the code less readable, as it detaches the contents of the function from the surrounding blocks that denote where the function starts and ends. a consistent policy is better than an inconsistent one, even if aesthetic judgements about it differ between contributors.

samoconnor commented on test/perf/kernel/stockcorr.jl:
```diff
@@ -4,7 +4,6 @@
 #STOCKCORR - The original, unoptimised code that simulates two correlated assets
 function stockcorr()
-
     ## Correlated asset information
```
Please don't delete blank lines preceding comments in cases like this where the author has clearly taken some time and care with manual ASCII layout to in an attempt to make the code easy to understand. A comment is the typographical equivalent of a heading, or the beginning of a paragraph, and should be clearly separated from whatever precedes it.

tkelman replied:
the extra line of separation here is completely redundant. the indentation already serves this exact purpose

tkelman committed:
NDRange does not exist any more

tkelman committed:
Remove pointless ... from end of comments

tkelman committed:
Remove some excess empty lines

2 in a row is generally enough

tkelman committed:
reindent test/char.jl

tkelman committed:
Fix output formatting and add doctests to strings manual

tkelman committed:
Make batchsplit errors more accurately reflect tested condition

would get MethodError or InexactError later for non integer inputs, but still

tkelman committed:
DRY some of the catch, retry, asyncmap and pmap tests in parallel_exec.jl

tkelman committed:
Remove many unnecessary end-of-line semicolons

tkelman committed:
Fix typo

tkelman committed:
Remove unecessary second names for Generator and AsyncGenerator

If these ever get exported, this can be reverted for consistency with
other exported iterator functions which are using a lower case convention
at the moment.

Leaving in place pgenerate since it at least has multiple nontrivial
signatures.
